// proto3記法であることを表している
syntax = "proto3";

package faq;

/*
   message命名規則(ベストプラクティスはどんなものだろう……)
 *Component : Request/Responseに用いるコンポーネント
 Faq[Action]Request : Action用リスエスト内容
 Faq[Action]Response : Action用レスポンス内容
*/

// FAQ
message FaqComponent {
    uint32 qid =1;
    message basic {
        uint32 share = 2;
        string service_name = 3;
    }

    message faq {
        string lang = 4;
        string question = 5;
        string answer = 6;
    }
}

// サーバーからのレスポンス(共通)
message ServerResponseComponent {
    bool is_success = 1;
    string message = 2;
}

// FAQを作成するときのリクエスト
message FaqCreateRequest {
    // この記載がダメなら、FaqComponentの内容を転記する
    repeated FaqComponent faqData = 1;
    string timestamp = 2;
}

// FAQを作成するときのレスポンス
// messageの中に別のmessageを含むことができる
message FaqCreateResponse {
    ServerResponseComponent response = 1;
    string timestamp = 2;
}

message FaqShowRequest {
    string timestamp = 1;
}

// repeated SOMETHING = SOMETHING[]
message FaqShowResponse {
    repeated FaqComponent faqs = 1;
    string timestamp = 2;
}

message FaqUpdateRequest {
    FaqComponent faq = 1;
    string timestamp = 2;
}

message FaqUpdateResponse {
    ServerResponseComponent response = 1;
    string timestamp = 2;
}

// ルーティング
service FaqGateway {
    rpc FaqCreate (FaqCreateRequest) returns (FaqCreateResponse) {}
    rpc FaqShow (FaqShowRequest) returns (FaqShowResponse) {};
    rpc FaqUpdate (FaqUpdateRequest) returns (FaqUpdateResponse) {}
}

